TestFoundation/TestURLCache.swift

```diff
@@ -0,0 +1,271 @@
+        
+        let urls = [ "https://apple.com/",
+                     "https://google.com/",
+                     "https://facebook.com/" ]
+        
+        var first = true
+        for (request, response) in try urls.map({ try cachePair(for: $0, ofSize: aBit) }) {
+            cache.storeCachedResponse(response, for: request)
+            if first {
+                Thread.sleep(forTimeInterval: 5.0)
+                first = false
+            }
+        }
+        
+        cache.removeCachedResponses(since: Date(timeIntervalSinceNow: -3.5))
+        
+        XCTAssertEqual(try FileManager.default.contentsOfDirectory(atPath: writableTestDirectoryURL.path).count, 1)
+        
+        first = true
+        for request in urls.map({ URLRequest(url: URL(string: $0)!) }) {
+            if first {
+                XCTAssertNotNil(cache.cachedResponse(for: request))
+            } else {
+                XCTAssertNil(cache.cachedResponse(for: request))
+            }
+            
+            first = false
+        }
+    }
+    
+    // -----
+    
+    static var allTests: [(String, (TestURLCache) -> () throws -> Void)] {
+        return [
+            ("testStorageRoundtrip", testStorageRoundtrip),
+            ("testStoragePolicy", testStoragePolicy),
+            ("testNoDiskUsageIfDisabled", testNoDiskUsageIfDisabled),
+            ("testShrinkingDiskCapacityEvictsItems", testShrinkingDiskCapacityEvictsItems),
+            ("testNoMemoryUsageIfDisabled", testNoMemoryUsageIfDisabled),
+            ("testShrinkingMemoryCapacityEvictsItems", testShrinkingMemoryCapacityEvictsItems),
+            ("testRemovingOne", testRemovingOne),
+            ("testRemovingAll", testRemovingAll),
+            ("testRemovingSince", testRemovingSince),
+        ]
+    }
+    
+    // -----
+    
+    func cache(memoryCapacity: Int = 0, diskCapacity: Int = 0) throws -> URLCache {
+        try FileManager.default.createDirectory(at: writableTestDirectoryURL, withIntermediateDirectories: true)
+        return URLCache(memoryCapacity: memoryCapacity, diskCapacity: diskCapacity, diskPath: writableTestDirectoryURL.path)
+    }
+    
+    func cachePair(for urlString: String, ofSize size: Int, storagePolicy: URLCache.StoragePolicy = .allowed) throws -> (URLRequest, CachedURLResponse) {
+        let url = try URL(string: urlString).unwrapped()
+        let request = URLRequest(url: url)
+        let response = try HTTPURLResponse(url: url, statusCode: 200, httpVersion: "1.1", headerFields: [:]).unwrapped()
+        return (request, CachedURLResponse(response: response, data: Data(count: size), storagePolicy: storagePolicy))
+    }
+    
+    var writableTestDirectoryURL: URL!
+    
+    override func setUp() {
+        super.setUp()
+        
+        let pid = ProcessInfo.processInfo.processIdentifier
+        writableTestDirectoryURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("org.swift.TestFoundation.TestURLCache.\(pid)")
```

Comment:
@drodriguez re: PID — it is understood that there's no change; the attempt here is to avoid a proliferation of temporary directories in case we have a bug with the logic here.

Reply:
I proposed it because using the same value in the different test might mean that a later test is affected by the previous one during the same run. It could lead to hard to debug problems where the test work (or doesn’t) when run in a suite, but fails (or works) when run individually.

It is not a big deal anyway.